### PR TITLE
interfaces: parallel instances support, extend unit tests

### PR DESCRIPTION
**Pull request by bboozzoo**
Allow adding slots/plugs from snaps with instance keys. Extend unit tests to
cover parallel instances of snaps.

### CODE REVIEW

File: interfaces/repo_test.go [hunk 1]
```diff
@@ -123,11 +127,18 @@ func (s *RepositorySuite) TearDownTest(c *C) {
 	s.BaseTest.TearDownTest(c)
 }
 
-func addPlugsSlots(c *C, repo *Repository, yamls ...string) []*snap.Info {
-	result := make([]*snap.Info, len(yamls))
-	for i, yaml := range yamls {
-		info := snaptest.MockInfo(c, yaml, nil)
-		result[i] = info
+func addPlugsSlotsFromInstances(c *C, repo *Repository, instanceNamesAndYamls ...string) []*snap.Info {
+	c.Assert(len(instanceNamesAndYamls)%2, Equals, 0, Commentf("unexpected length"))
```
Comment (stolowski) [inline interfaces/repo_test.go]: "unexpected length: an even number expected" or "unexpected length: expected instance name, yaml pairs" would make the intention clear IMHO.

File: interfaces/repo_test.go [hunk 2]
```diff
@@ -1993,6 +2127,23 @@ slots:
   i2:
 `), nil)
 	c.Assert(r.AddSnap(s3), IsNil)
+	s3Instance := snaptest.MockInfo(c, fmt.Sprintf(`
```
Comment (stolowski) [inline interfaces/repo_test.go]: I think `fmt.Sprintf` is not needed here, is it?

File: interfaces/repo_test.go [hunk 3]
```diff
@@ -1993,6 +2127,23 @@ slots:
...
+name: s3
+version: 0
+type: os
+slots:
+  i2:
+`), nil)
+	s3Instance.InstanceKey = "instance"
+	c.Assert(r.AddSnap(s3Instance), IsNil)
+	s4 := snaptest.MockInfo(c, fmt.Sprintf(`
```
Comment (stolowski) [inline interfaces/repo_test.go]: Same.

File: interfaces/sorting.go [hunk 4]
```diff
@@ -69,21 +69,27 @@ type byPlugSnapAndName []*snap.PlugInfo
...
-	if c[i].Snap.InstanceName() != c[j].Snap.InstanceName() {
-		return c[i].Snap.InstanceName() < c[j].Snap.InstanceName()
+	if c[i].Snap.SnapName() != c[j].Snap.SnapName() {
+		return c[i].Snap.SnapName() < c[j].Snap.SnapName()
 	}
-	return c[i].Name < c[j].Name
+	if c[i].Name != c[j].Name {
+		return c[i].Name < c[j].Name
+	}
+	return c[i].Snap.InstanceKey < c[j].Snap.InstanceKey
```
Comment (stolowski) [inline interfaces/sorting.go]: Shouldn't we sort by snap name, then instance key and then plug name to be in line with what we did before instances were introduced?

File: interfaces/sorting.go [hunk 5]
```diff
@@ -69,21 +69,27 @@ type byPlugSnapAndName []*snap.PlugInfo
...
```
Comment (stolowski) [inline interfaces/sorting.go]: Same question.

File: interfaces/sorting.go [hunk 6]
```diff
@@ -143,19 +149,25 @@ type byPlugInfo []*snap.PlugInfo
...
```
Comment (stolowski) [inline interfaces/sorting.go]: Same question re sort order.

File: interfaces/sorting.go [hunk 7]
```diff
@@ -143,19 +149,25 @@ type byPlugInfo []*snap.PlugInfo
...
```
Comment (stolowski) [inline interfaces/sorting.go]: Same.

File: interfaces/sorting.go (hunk 4 above)
Comment (bboozzoo) [reply interfaces/sorting.go]: Current code orders like this: `foo:a`, `foo_b:a`, `foo:b`. Your proposal would order like this: `foo:a`, `foo:b`, `foo_b:a`. Either works for me, though I find the first variant bit easier to reason about.

Comment (bboozzoo) [reply interfaces/sorting.go]: Now that I read it again I see it makes sense. `foo` & `foo_bar` are different snaps after all.

File: interfaces/repo_test.go [hunk 8]
```diff
@@ -307,6 +318,21 @@ func (s *RepositorySuite) TestAddPlugFailsWithUnknownInterface(c *C) {
 	c.Assert(s.emptyRepo.AllPlugs(""), HasLen, 0)
 }
 
+func (s *RepositorySuite) TestAddPlugParallelInstance(c *C) {
+	c.Assert(s.testRepo.AllPlugs(""), HasLen, 0)
+	err := s.testRepo.AddPlug(s.plug)
+	c.Assert(err, IsNil)
+	consumer := snaptest.MockInfo(c, consumerYaml, nil)
```
Comment (mvo5) [inline interfaces/repo_test.go]: I think I would find it slightly clearer when `c.Assert(s.testRepo.AllPlugs(""), HasLen, 1)` was added here. It just feels more natural to me to not jump from zero to two.

I.e.:
```
	c.Assert(s.testRepo.AllPlugs(""), HasLen, 0)
	err := s.testRepo.AddPlug(s.plug)
	c.Assert(err, IsNil)
	c.Assert(s.testRepo.AllPlugs(""), HasLen, 1)

	consumer := snaptest.MockInfo(c, consumerYaml, nil)
	consumer.InstanceKey = "instance"
	err = s.testRepo.AddPlug(consumer.Plugs["plug"])
	c.Assert(err, IsNil)
	c.Assert(s.testRepo.AllPlugs(""), HasLen, 2)
        c.Assert(s.testRepo.Plug(s.plug.Snap.InstanceName(), s.plug.Name), DeepEquals, s.plug)
	c.Assert(s.testRepo.Plug(consumer.InstanceName(), "plug"), DeepEquals, consumer.Plugs["plug"])
```

File: interfaces/repo_test.go [hunk 9]
```diff
@@ -317,16 +343,23 @@ func (s *RepositorySuite) TestPlug(c *C) {
 }
 
 func (s *RepositorySuite) TestPlugSearch(c *C) {
-	addPlugsSlots(c, s.testRepo, `
+	addPlugsSlotsFromInstances(c, s.testRepo, "x", `
```
Comment (mvo5) [inline interfaces/repo_test.go]: I find this hard to read. Maybe:
```
type instanceNameAndYaml struct {
	Name string
	Yaml string
}

func addPlugsSlotsFromInstances(c *C, repo *Repository, iys []instanceNameAndYaml) []*snap.Info {
...
```
The full diff: http://paste.ubuntu.com/p/RBG9C3MGws/

File: interfaces/repo_test.go [hunk 10]
```diff
@@ -582,6 +673,21 @@ func (s *RepositorySuite) TestAddSlotStoresCorrectData(c *C) {
 	c.Assert(slot, DeepEquals, s.slot)
 }
 
+func (s *RepositorySuite) TestAddSlotParallelInstance(c *C) {
+	c.Assert(s.testRepo.AllSlots(""), HasLen, 0)
+	err := s.testRepo.AddSlot(s.slot)
+	c.Assert(err, IsNil)
```
Comment (mvo5) [inline interfaces/repo_test.go]: Would love to have a "c.Assert(s.testRepo.AllSlots(""), HasLen, 1)" here too (same as above)

File: interfaces/repo_test.go (hunk 9 above)
Comment (bboozzoo) [reply interfaces/repo_test.go]: Thank you!. Applied the diff with some minor tweaks.

**Pull request merged**